compiler/luci/pass/src/FoldShapePass.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/FoldShapePass.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Profile/CircleNodeOrigin.h>
+
+namespace
+{
+
+template <loco::DataType OutType> luci::CircleConst *folding(luci::CircleShape *shape)
+{
+  auto input_node = loco::must_cast<luci::CircleNode *>(shape->input());
+  auto name = input_node->name();
+  assert(name.length() > 0);
+  auto shape_status = input_node->shape_status();
+  if (shape_status != luci::ShapeStatus::VALID)
+    return nullptr;
+  auto rank = input_node->rank();
+  if (rank == 0)
+    return nullptr;
+  for (uint32_t i = 0; i < rank; i++)
+  {
+    auto dim = input_node->dim(i);
+    if (!dim.known())
+      return nullptr;
+  }
+
+  auto folded_shape = input_node->graph()->nodes()->create<luci::CircleConst>();
+  folded_shape->name(name + "_ConstShape");
+  folded_shape->dtype(OutType);
+  folded_shape->rank(1);
+  folded_shape->dim(0).set(rank);
+  luci::add_origin(folded_shape, luci::get_origin(shape));
+
+  folded_shape->size<OutType>(rank);
+  for (uint32_t i = 0; i < rank; i++)
+    folded_shape->at<OutType>(i) = input_node->dim(i).value();
+
+  return folded_shape;
+}
+
+// Fold Shape to const if the input shape is static
+template <loco::DataType OutType> bool fold_shape(luci::CircleShape *shape)
+{
+  auto folded_shape = folding<OutType>(shape);
+  if (not folded_shape)
+    return false;
+
+  loco::replace(shape).with(folded_shape);
+
+  return true;
+}
+
+} // namespace
+
+namespace luci
+{
+
+/**
+ * BEFORE
+ *
+ *        [Input]
+ *           |
+ *        [Shape]
+ *           |
+ *        [Output]
+ *
+ * AFTER
+ *
+ *      [CircleConst]
+ *
```

Comment:
where did `[Input]` and `[Output]` go?
Q) what do you mean by `[Input]` and `[Output]` ? or, what are these?

Reply:
> Q) what do you mean by `[Input]` and `[Output]` ? or, what are these?

I wanted to say that there is a tensor as an `[Input]` of `Shape` (that has a static shape... which can be folded as a constant), and then it returns a tensor as an `[Output]` (rank 1 tensor whose values are the each dimension of input tensor)

And after the `Shape` is folded as a constant, the `[Output]` will be replaced as a `[CircleConst]`

This is what I wanted to say... but the diagram seems insufficient to explain it 🤔


Is it more proper to comment like below..?

```cpp
 * BEFORE
 *
 *           |
 *        [Shape]
 *           |
 *
 * AFTER
 *
 *      [CircleConst]
```

`|` above the `[Shape]` is an input tensor, and `|` below the `[Shape]` is an output tensor

Reply:
`[CircleSomething]` in the diagram is intended to express the circle node IR.
If you used `[CircleConst]` then it would be the `CircleConst` IR node.
Using `[Shape]`, not `[CircleShape]` would not follow this and also 
you have two different expression rules in your diagram.

Using `[CircleNode]` is used as don't care about specific IR but it is a circle IR node.

I think you may have copied `[Input]` and `[Output]` from some where else...
In circle IR, there is `CircleInput` and `CircleOutput` to express inputs and outputs of the graph. I was curious if you knew this when adding the `[Input]` and `[Output]` in the graph.

> where did [Input] and [Output] go?

This question is about why are replaced nodes not in the diagram.
To be exact, all the pass don't remove the node, just changes the connection.
If you look in to some of the graphs, but not all, there are diagrams that shows dangling replaced nodes. This is to express nature of the Pass. Real deletion is done in some other Pass, for that purpose.

Reply:
First, I truley appreciate to your kind response :)

>In circle IR, there is `CircleInput` and `CircleOutput` to express inputs and outputs of the graph. I was curious if you knew this when adding the `[Input]` and `[Output]` in the graph.

I didn't know... I just thought that the `Input`/`Output` means `before`/`after` node of `Shape` operation respectively (Input of `Shape` and output of `Shape`)

> To be exact, all the pass don't remove the node, just changes the connection.

Then I guess, according to this, below diagram is more exact to explain the behavior of this pass. Am I right..?

```cpp
 * BEFORE
 * 
 *     [CircleInput]
 *           |
 *        [Shape]
 *           |
 *     [CircleOutput]
 *
 * AFTER
 *
 *     [CircleInput]
 *
 *     [CircleConst]
 *           |
 *     [CircleOutput]
 *
```

---
> If you look in to some of the graphs, but not all, there are diagrams that shows dangling replaced nodes.

Yes, I can find it in `FuseAddWithTConvPass.cpp`

Reply:
If this `CircleInput` or `CircleOutput` is not a requirement, it would be better to use just `CircleNode` as these can be any kind of circle nodes.